### PR TITLE
SDK-812: Added interactive workflow transition resolver.

### PR DESCRIPTION
**Pull request by vol4onok**
https://spryker.atlassian.net/browse/SDK-812

### CODE REVIEW

**Review by sergeyspryker — APPROVED** [review]

**Review by Kyrylo-Meretskyi — APPROVED** [review]

File: src/Extension/Workflow/InteractionAnswerBasedTransitionResolver.php [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+{
+    /**
+     * @var string
+     */
+    protected const RESOLVER_NAME = 'interactive';
+
+    /**
+     * @var string
+     */
+    public const QUESTION = 'question';
```
Comment (pavelmaksimov25) [inline src/Extension/Workflow/InteractionAnswerBasedTransitionResolver.php:26]: Public const should be above protected

File: tests/Sdk/Unit/Extension/Workflow/InteractionAnswerBasedTransitionResolverTest.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+namespace SprykerSdk\Sdk\Unit\Extension\Workflow;
+
+use Codeception\Test\Unit;
+use SprykerSdk\Sdk\Extension\Workflow\InteractionAnswerBasedTransitionResolver;
+use SprykerSdk\SdkContracts\Entity\ContextInterface;
+use SprykerSdk\SdkContracts\ValueReceiver\ValueReceiverInterface;
+
+class InteractionAnswerBasedTransitionResolverTest extends Unit
```
Comment (pavelmaksimov25) [inline tests/Sdk/Unit/Extension/Workflow/InteractionAnswerBasedTransitionResolverTest.php:15]: Please provide test that checks if exception is thrown if it is a valid case.

File: src/Extension/Workflow/InteractionAnswerBasedTransitionResolver.php [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+     */
+    protected function getChoiceValues(array $settings): array
+    {
+        $choices = $settings[static::CHOICES] ?? [];
+        $choiceValues = [];
+        foreach ($choices as $transition => $choice) {
+            $choiceValues[(string)$transition] = (string)$choice['description'];
+        }
+
+        $duplicates = array_unique(array_diff_assoc($choiceValues, array_unique($choiceValues)));
```
Comment (pavelmaksimov25) [inline src/Extension/Workflow/InteractionAnswerBasedTransitionResolver.php:90]: This piece of logic is a bit complicated. Please avoid nested expressions to decrease complexity and provide more readability.
Also, it is not clear, why we shouldn't check description uniqueness. IMO, having unique keys is more important check

Comment (pavelmaksimov25) [reply src/Extension/Workflow/InteractionAnswerBasedTransitionResolver.php:90]: IMO `count($choiceValues) === count(array_unique($choiceValues))` looks simplier.

File: extension/Hello/config/workflow.yaml [hunk 4]
```diff
@@ -53,6 +53,20 @@ framework:
               settings:
                 failed: bye # transition name for filed result
                 successful: world # transition name for successful result
+        hello-interactive:
+          from: start
+          to: middle
+          metadata:
+            task: ~ # not neccessary to have a task for the interactive transition.
+            transitionResolver: # Resolver needs for resolve next transition
+              name: interactive
```
Comment (pavelmaksimov25) [inline extension/Hello/config/workflow.yaml:62]: please add a comment "By setting `interactive` transitionResolver name current transition will be recognized as interactive and appropriate actions will be done."

**Review by pavelmaksimov25 — CHANGES_REQUESTED** [review]
Please merge the latest master and check how it works with php v 7.4
Also, we need to update our `docs/workflow.md` file with the additional block that describes how interactive workflow works. Just a few sentences should be enough.

File: src/Extension/Workflow/InteractionAnswerBasedTransitionResolver.php (hunk 1 above)
Comment (vol4onok) [reply src/Extension/Workflow/InteractionAnswerBasedTransitionResolver.php:26]: Why? It's keys for structure I use it outside of the class

File: src/Extension/Workflow/InteractionAnswerBasedTransitionResolver.php (hunk 3 above)
Comment (vol4onok) [reply src/Extension/Workflow/InteractionAnswerBasedTransitionResolver.php:90]: it's not enough to check. I need to show these duplicates

File: tests/Sdk/Unit/Extension/Workflow/InteractionAnswerBasedTransitionResolverTest.php (hunk 2 above)
Comment (vol4onok) [reply tests/Sdk/Unit/Extension/Workflow/InteractionAnswerBasedTransitionResolverTest.php:16]: done

File: src/Extension/Workflow/InteractionAnswerBasedTransitionResolver.php (hunk 1 above)
Comment (pavelmaksimov25) [reply src/Extension/Workflow/InteractionAnswerBasedTransitionResolver.php:26]: it is a common structure for class - public -> protected -> private. Please put public above protected

**Review by pavelmaksimov25 — CHANGES_REQUESTED** [review]

File: src/Extension/Workflow/InteractionAnswerBasedTransitionResolver.php (hunk 3 above)
Comment (pavelmaksimov25) [reply src/Extension/Workflow/InteractionAnswerBasedTransitionResolver.php]: ok. but do we really need to control description uniqueness?

**Review by pavelmaksimov25 — DISMISSED** [review]

File: extension/Hello/config/workflow.yaml [hunk 5]
```diff
@@ -53,6 +53,20 @@ framework:
               settings:
                 failed: bye # transition name for filed result
                 successful: world # transition name for successful result
+        hello-interactive:
```
Comment (Kyrylo-Meretskyi) [inline extension/Hello/config/workflow.yaml:56]: Fix indentation, because this workflow isn't shown in workflow list

**Review by Kyrylo-Meretskyi — CHANGES_REQUESTED** [review]

**Review by pavelmaksimov25 — APPROVED** [review]

**Review by sergeyspryker — APPROVED** [review]

**Pull request merged**